File: valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    
    let map = {
        '(': ')',
        '{': '}',
        '[': ']'
    }
    
    for(let char of s) {
        if(char in map) {
            stack.push(char);
        } else {
            let pop = stack.splice(-1);
            if (map[pop] !== char) {
                return false;
            }
        }
    }
    if (stack.length !== 0) return false;
    return true;
};

File: valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        // List<Integer> list=new ArrayList<Integer>();
        // List<Integer> llist=new LinkedList<Integer>();
        Stack<Character> stack = new Stack<Character>();
        HashMap<Character, Character> parens = new HashMap<Character, Character>();
        
        parens.put('(', ')');
        parens.put('[', ']');
        parens.put('{', '}');
        
        for(int i = 0; i < s.length(); i++) {
            if(parens.containsKey(s.charAt(i))) {
                stack.add(s.charAt(i));
            } else {
                if(stack.isEmpty()) {
                    return false;
                } else if (s.charAt(i) != parens.get(stack.pop())){
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}